### ISSUE TITLE
UTF-8 manpages

### ISSUE THREAD

**Issue by edwintorok** [dbuenzli/cmdliner#27]
"However at the moment Groff (at least 1.19.2) doesn't seem to cope with UTF-8 input and UTF-8 characters beyond the ASCII set will look garbled."

I found a workaround: use `preconv(1)` or `groff -k -man -Tutf8`.
Suggestion:
1. Change `Pipe this output to groff -man -Tutf8 | less` to `Pipe this output to groff  -K utf8 -man -Tutf8 | less -s`
2. When automatically running groff, use `-K utf8`
3. `--help=groff` could perform the conversion done by `preconv(1)` on its own, so no `-K utf8` is needed anymore

For example you can currently do this on Debian jessie to get the UTF8 printed correctly:
```sh
./noop --help=groff | groff -k -man -Tutf8 | pager -s
./noop --help=groff | preconv -e utf-8 >noop.1 && groff -man -Tutf8 <noop.1 | pager -s
```

It looks like this happens automatically when you call `man noop` in an UTF-8 locale
(and in a C locale the unicode characters get transliterated to ASCII):
```sh
sudo cp noop.1 /usr/local/man/man1/
man noop
LC_ALL=C man noop
```

The manpage `noop.1` displays even on OpenBSD though not with default flags (by default it replaces all UTF-8 characters with ?):
```sh
mandoc -Tutf8 noop.1 | LC_CTYPE=C.UTF-8 more -s
```

I haven't experienced this behaviour though: "-Tutf8 maps the hyphen-minus U+002D to the minus sign U+2212", and `-Tutf8` is require to get the UTF-8 chars printed at all.

**Comment by dbuenzli** [comment]
Sadly this doesn't seem to work on osx.

**Comment by dbuenzli** [comment]
If I understand correctly `preconv` simply preprocesses the UTF-8 input to `\[uXXXX]`. That's not rocket science but I'm neither willing to add an `uutf` dep on `cmdliner` nor write yet another UTF-8 decoding function.

It's a bit sad `osx` is lagging so much with `groff`. `preconv` was [introduced](http://git.savannah.gnu.org/cgit/groff.git/tree/NEWS#n237) in `groff` 1.20, osx 10.12 has 1.19.2.

**Comment by dbuenzli** [comment]
Prompted by @MisterDA I just had a quick look again at that. 

On macOS 11 the version of groff is still stuck at 1.19.2. However I tested this invocation: 

```
mandoc noop.1 | less
```

and it works (with `-Tutf8` aswell) and the `-` are minuses. So maybe we could switch to that in the [paging code](https://github.com/dbuenzli/cmdliner/blob/db4d02a9eb47b5c43127a67cb121004b03ea3719/src/cmdliner_manpage.ml#L443-L470).

But has anyone has a good idea on how `mandoc` relates to `groff`  ? Is `mandoc` readily available on distributions ? Do we have a good fallback if not ?

**Comment by dbuenzli** [comment]
Also is `mandoc` the package different from `mandoc` the tool ? For example this doesn't work:

```
groff -mandoc -Tutf8 noop.1
```

**Comment by MisterDA** [comment]
IIRC mandoc-the-tool parses mdoc-the-format, and is used on BSDs for the man pages. It's available in most Linux distros, sometimes under its old name of mdocml (see [repology](https://repology.org/project/mandoc/versions)) but not used much by default to the best of my knowledge.

**Comment by dbuenzli** [comment]
> IIRC mandoc-the-tool parses mdoc-the-format, and is used on BSDs for the man pages.

The next question is how the mdoc-the-format relates to what cmdliner generate :-) 

This may turn out a bit brittle but maybe we could try to prepend `mandoc` to the fallback (this will catch `macOS`) and then invoke `groff` with `-k` which is unsupported on `macOS` but it [looks](https://repology.org/project/groff/versions) well supported on linux distros (needs at least groff 1.20)

This means try in order: 
```
mandoc    # invoke mandoc -Tutf8 noop.1
groff     # invoke groff -k -Tutf8 noop.1
nroff     # invoke nroff noop.1
```

Also I don't remember what the `-P-c` is for on [this line](https://github.com/dbuenzli/cmdliner/blob/db4d02a9eb47b5c43127a67cb121004b03ea3719/src/cmdliner_manpage.ml#L464).

We would also need confirmation that this works for the other BSDers.

**Comment by MisterDA** [comment]
Is there a reason not to use also `groff -Kutf8` to specify that the input is utf8? if of course, that's the encoding cmdliner is producing. preconv has a mechanism to detect the encoding, but if we know it beforehand we might as well tell it.

> The next question is how the mdoc-the-format relates to what cmdliner generate :-)

There are two formats: man, historical one, used on most Linux distros, and mdoc, used on BSDs (appeared in 4.4BSD, 1997). GNU troff (groff) and the BSD mandoc are able to interpret both formats, and both can detect what format is used. groff has good support for mdoc since 1.17 (2001). There is a list of compatibility issues between mandoc and groff support of mdoc. https://mandoc.bsd.lv/man/mdoc.7.html#COMPATIBILITY

> Also I don't remember what the -P-c is for on this line.

>-P <ins>-option</ins>
>-P <ins>-option</ins> -P <ins>arg</ins>
>
> Pass <ins>-option</ins> or <ins>-option</ins> arg to the postprocessor. The option must be specified with the necessary preceding minus sign(s) ‘-’ or ‘--’ because *groff* does not prepend any dashes before passing it to the postprocessor. For example, to pass a title to the *gxditview* postprocessor, the shell command
>
>     groff -X -P -title -P 'groff it' foo
>
>  is equivalent to
>
>     groff -X -Z foo | gxditview -title 'groff it' -
>
>-c
>
> Disable color output. Please consult the [grotty(1)][grotty.1] man page for more details.

[grotty.1]: https://manpages.ubuntu.com/manpages/precise/en/man1/grotty.1.html

https://explainshell.com/explain?cmd=groff+-T+utf8+-P-c

So now we know *what* `-P-c` does. Without the `-c` switch, grotty outputs [SGR parameters](https://en.wikipedia.org/wiki/ANSI_escape_code#SGR_(Select_Graphic_Rendition)_parameters), but this requires support from the pager. The manual page suggests parameters to the pager in its manual, namely `less -U` or `less -B` or `more -b` when using `grotty -c`, and `less -R` with the new output format (without grotty `-c` flag).

**Comment by dbuenzli** [comment]
Thanks @MisterDA for the help. 

I have to admit I evacuated all these man idosyncrasies out of my brain – that's one of the reason why cmdliner exists!

> Is there a reason not to use also `groff -Kutf8` to specify that the input is utf8?

No, I'm fine with telling people cmdliner doc string have to be UTF-8 encoded.

@MisterDA I will likely have a look at these things when I switch back to `cmdliner` before the end of this year. But if you have time to provide us with the perfect multiplatform logic for UTF-8 and whatever coloring before, it's just a matter of tweaking [this function](https://github.com/dbuenzli/cmdliner/blob/db4d02a9eb47b5c43127a67cb121004b03ea3719/src/cmdliner_manpage.ml#L443-L470).

**Comment by dbuenzli** [comment]
Thanks to @MisterDA's excellent dive into the problem in 2100285dcec084fe9b4040 and 3fe9fcf5ea991bb3bf37117. It seems we have something that works on linux and macos. 

One slight caveat is that on macos it doesn't work if you install the manpages output via `--help=groff` but I'm willing to pass over that.

We need to confirm that it works on the BSDs, here are instruction 

```
opam install topkg-care
git clone https://erratique.ch/repos/cmdliner.git
cd cmdliner
topkg build 
topkg run -- test/test_man_utf8 --help
```
The characters in the synopsis of the man page should match those [here](https://github.com/dbuenzli/cmdliner/blob/fee08709a44af4340d84b651de9f1e52377dff3d/test/test_man_utf8.ml#L9). 

Maybe @hannesm ?

**Comment by dbuenzli** [comment]
Thanks, Hannes !

**Comment by dbuenzli** [comment]
Thanks ! Do you have `mandoc` in your `PATH` ? 

Actually I'm starting to think that maybe I should bite the bullet and implement [this](https://github.com/dbuenzli/cmdliner/issues/27#issuecomment-946253506).

**Comment by MisterDA** [comment]
> @MisterDA are you talking about this commit ?

No, sorry if that wasn't clear. I'm talking about the `FMT` argument to the default `--help` option:

> ```
>        --help[=FMT] (default=auto)
>            Show this help in format FMT. The value FMT must be one of `auto',
>            `pager', `groff' or `plain'. With `auto', the format is `pager` or
>            `plain' whenever the TERM env var is `dumb' or undefined.
> ```

I'm suggesting to replace the `groff` value of the enumeration by `man`. If cmdliner supports `mandoc`, I think it's better to indicate the format cmdliner is outputting instead of a specific tool that understands it.
